apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts

```diff
@@ -0,0 +1,1029 @@
+                  }
+                })
+            );
+          };
+        }
+
+        for (const sourceFile of tsProgram.getSourceFiles()) {
+          const filename: string = sourceFile.fileName;
+          if (typescriptFilenames.has(filename)) {
+            const relativeFilenameWithoutExtension: string = ts.removeFileExtension(
+              ts.getExternalModuleNameFromPath(resolverHost, filename)
+            );
+
+            for (const { cacheOutFolderPath, outFolderPath, isPrimary } of this._moduleKindsToEmit) {
+              if (isPrimary) {
+                // Only primary module kinds emit declarations and sourcemaps
+                if (tsconfig.options.declaration) {
+                  const dtsFilename: string = `${relativeFilenameWithoutExtension}.d.ts`;
+                  queueLinkOrCopy({
+                    linkTargetPath: path.join(cacheOutFolderPath, dtsFilename),
+                    newLinkPath: path.join(outFolderPath, dtsFilename)
+                  });
+                }
+
+                if (tsconfig.options.sourceMap && !sourceFile.isDeclarationFile) {
+                  const jsMapFilename: string = `${relativeFilenameWithoutExtension}.js.map`;
+                  queueLinkOrCopy({
+                    linkTargetPath: path.join(cacheOutFolderPath, jsMapFilename),
+                    newLinkPath: path.join(outFolderPath, jsMapFilename)
+                  });
+                }
+
+                if (tsconfig.options.declarationMap) {
+                  const dtsMapFilename: string = `${relativeFilenameWithoutExtension}.d.ts.map`;
+                  queueLinkOrCopy({
+                    linkTargetPath: path.join(cacheOutFolderPath, dtsMapFilename),
+                    newLinkPath: path.join(outFolderPath, dtsMapFilename)
+                  });
+                }
+              }
+
+              if (!sourceFile.isDeclarationFile) {
+                const jsFilename: string = `${relativeFilenameWithoutExtension}.js`;
+                queueLinkOrCopy({
+                  linkTargetPath: path.join(cacheOutFolderPath, jsFilename),
+                  newLinkPath: path.join(outFolderPath, jsFilename)
+                });
+              }
+            }
+          }
+        }
+
+        await Promise.all(linkPromises);
+
+        return { linkCount };
+      }
+    );
+
+    this._terminal.writeVerboseLine(
+      `${shouldHardlink ? 'Hardlink' : 'Copy from cache'}: ${hardlinkDuration}ms (${hardlinkCount} files)`
+    );
+    //#endregion
+
+    let compilationFailed: boolean = false;
+    if (diagnostics.length > 0) {
+      this._terminal.writeErrorLine(
+        `Encountered ${diagnostics.length} TypeScript error${diagnostics.length > 1 ? 's' : ''}:`
+      );
+      for (const diagnostic of diagnostics) {
+        this._printDiagnosticMessage(ts, diagnostic, true);
+      }
+
+      compilationFailed = true;
+    }
+
+    if (tslintResult?.failures?.length) {
+      this._terminal.writeWarningLine(
```

Comment:
Continuing my example from above, the log output is colored as follows:

In red color:
```
[typescript] Encountered 3 TypeScript errors:
[typescript]   ERROR: src\index.ts(5,7) - (TS6133) 'x' is declared but its value is never read.
[typescript]   ERROR: src\index.ts(5,15) - (TS2304) Cannot find name 'ForgottenImport'.
[typescript]   ERROR: src\index.ts(6,7) - (TS6133) 'p' is declared but its value is never read.
```

In yellow color:
```
[typescript] Encountered 5 TSLint errors:
[typescript]   ERROR: src\index.ts(6,29) - (no-empty) block is empty
[typescript]   ERROR: src\index.ts(5,7) - (prefer-const) Identifier 'x' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(6,7) - (prefer-const) Identifier 'p' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(5,7) - (typedef) expected variable-declaration: 'x' to have a typedef
[typescript]   ERROR: src\index.ts(6,7) - (typedef) expected variable-declaration: 'p' to have a typedef
```

Questions:
- Why are warnings referred to as "errors"?
- Why are TSLint problems prefixed with `[typescript]`?

There are probably nerdy excuses, but as a user experience it seems not right.

Reply:
I've given the tslint logging messages their own prefix for now. I think we need to have a longer design discussion about this.